stunner/stunner.go

```diff
@@ -228,29 +253,36 @@ func (s *Stunner) sendSTUN(ctx context.Context, server string) error {
 	}
 	if addr.IP == nil {
 		if s.OnlyIPv6 {
-			return fmt.Errorf("cannot resolve any ipv6 addresses for %s, got: %v", server, ipAddrs)
+			return nil, fmt.Errorf("cannot resolve any ipv6 addresses for %s, got: %v", server, ipAddrs)
 		}
-		return fmt.Errorf("cannot resolve any ipv4 addresses for %s, got: %v", server, ipAddrs)
+		return nil, fmt.Errorf("cannot resolve any ipv4 addresses for %s, got: %v", server, ipAddrs)
 	}
+	return addr, nil
+}
 
-	txID := stun.NewTxID()
-	req := stun.Request(txID)
-	s.addTX(txID, server)
-	_, err = s.Send(req, addr)
+func (s *Stunner) sendPackets(ctx context.Context, server string) error {
+	addr, err := s.serverAddr(ctx, server)
 	if err != nil {
-		return fmt.Errorf("send: %v", err)
+		return err
 	}
-	return nil
-}
 
-var retryDurations = []time.Duration{
-	100 * time.Millisecond,
-	100 * time.Millisecond,
-	100 * time.Millisecond,
-	200 * time.Millisecond,
-	200 * time.Millisecond,
-	400 * time.Millisecond,
-	800 * time.Millisecond,
-	1600 * time.Millisecond,
-	3200 * time.Millisecond,
+	const maxSend = 2
+	for i := 0; i < maxSend; i++ {
+		txID := stun.NewTxID()
+		req := stun.Request(txID)
+		s.addTX(txID, server)
+		_, err = s.Send(req, addr)
+		if err != nil {
+			return fmt.Errorf("send: %v", err)
+		}
+
+		select {
+		case <-ctx.Done():
+			// Ignore error. The caller deals with handling contexts.
+			// We only use it to dermine when to stop spraying STUN packets.
+			break
+		case <-time.After(time.Millisecond * time.Duration(50+rand.Intn(200))):
```

Comment:
At least half the DERP servers will take 100-200ms to respond, so this will always retry at least once. Still a bit aggressive?

Reply:
I ended this change a bit early because it was getting big and I got to a good stopping point. The further plan is to collect a couple STUN pings from each node (to help with #162), as long as we can within an overall limit (like 750ms or 1ms or the time to collect a minimum 1 STUN ping from each node).

And then we'll limit the number of STUN pings for the DERPs likely to still be far away. So we'll only be more paranoid-ly timing our close ones (say, NY and SF if you're in the US). But we do want multiple in the general case. It's just that I'm currently throwing those away. Later I'll both keep those, and have netcheck keep a few history netcheck.Report records.